### PR TITLE
Use ddev.site instead of ddev.local for site domains, provide offline fallback for hosts entry

### PR DESCRIPTION
**Pull request by tannerjfco**
## The Problem/Issue/Bug:
#416 OP - We want to provide a real domain for local dev sites to use in order to avoid hosts file modification. We still want to provide the hosts modification as a fallback behavior.

## How this PR Solves The Problem:

- Changes local TLD from ddev.local to ddev.site, which is configured to route to 127.0.0.1 for any requests.
- Introduces a new offline mode, which is controlled with the new `ddev offline` command. I admit I jumped the gun on this part, as its proposed but not yet accepted, but I think its a good way to handle the concerns we had switching to use ddev.site.

## Manual Testing Instructions:

- configure a new ddev site and start it. The URL for the site should be <sitename>.ddev.site instead of <sitename>.ddev.local. You should not be prompted to create a hosts entry
- run `ddev offline enable`. You should now be prompted to create a hosts entry for the site. For extra credit,  you can spin up another site or two before enabling offline, and see that any running sites should be part of the hosts entry being created.
- configure another ddev site and start it. You should be prompted to create a hosts entry for the site.
- running `ddev offline disable` should make it so any new site spin ups do not prompt for hosts entry

### CODE REVIEW

File: pkg/plugins/platform/local.go [hunk 1]
```diff
@@ -898,7 +895,7 @@ func (l *LocalApp) AddHostsEntry() error {
 	util.CheckErr(err)
 
 	fmt.Println("ddev needs to add an entry to your hostfile.\nIt will require root privileges via the sudo command, so you may be required\nto enter your password for sudo. ddev is about to issue the command:")
-	hostnameArgs := []string{ddevFullpath, "hostname", l.HostName(), dockerIP}
+	hostnameArgs := []string{ddevFullpath, "hostname", dockerIP, l.HostName()}
```
Comment (rfay) [inline pkg/plugins/platform/local.go]: I understand that this corresponds to the order in /etc/hosts but I think it was easier to use the other way.

File: pkg/plugins/platform/utils.go [hunk 2]
```diff
@@ -208,3 +214,79 @@ func ddevContainersRunning() (bool, error) {
 	}
 	return false, nil
 }
+
+// SetOfflineMode enables offline mode for ddev. When offline mode is
+// set, a tracking file is created and hosts entries are created for
```
Comment (rfay) [inline pkg/plugins/platform/utils.go]: Say ~/.ddev/.offline instead of "a tracking file"

Comment (rfay) [reply pkg/plugins/platform/utils.go]: Should we reverse the logic and make it offline by default? So maybe make a file named .usedns?

File: pkg/plugins/platform/utils.go [hunk 3]
```diff
@@ -208,3 +214,79 @@ func ddevContainersRunning() (bool, error) {
...
+// UnsetOfflineMode removes the tracking file for offline mode in order
+// to disable it.
+func UnsetOfflineMode() error {
+	err := os.Remove(offlineFile)
+	return err
+}
+
+func getDockerIP() string {
+	dockerIP := "127.0.0.1"
+	dockerHostRawURL := os.Getenv("DOCKER_HOST")
```
Comment (rfay) [inline pkg/plugins/platform/utils.go]: Thanks for this!

**Comment by rickmanelius** [comment]
I'm running out of time and battery on the plane, so I'll save the lengthier point by point review and testing. However, I've given this a spin and I'm ready to unblock this. Comments:

* Not sure if it's related to anything in this, but while I was able to get a D7.56 to spin up, it 500/501s. WordPress 4.9 was completely fine.
* I don't know why "offline" seems like an odd word to me. I understand functionally it's more a delination that I need to be online for the non /etc/host modification requirement. I can't think of a better word off the top of my head, so I won't let that be a deal breaker.
* While @rfay mentioned that perhaps we reverse the logic by default, I think we should default to online mode because one of the big drivers for this was not requiring the security escalation for the /etc/hosts modification. That said, there probably has to be *some* warning or help if a user happens to have no wifi and the site is hanging. This would reduce support requests for people tripping up on this. Additionally, if/when we provide the ability for an end-user to specify a custom domain, this type of checking and /etc/host modification is basically a necessity. From that perspective, I could see driving back to offline by default, but I think we'll ahve to wait for feedback to know which approach is the majority use case.
* Overall this worked as intended. I also appreciate the care put into the helper text. I haven't reviewed thoroughly, but that doesn't block the functionality and is small stuff that could change later on.

TL;DR, I'm 👍 with this as an end-user UX and I believe this balances the needs we've discussed. If @rfay is good on the code review side, let's proceed.

**Comment by cweagans** [comment]
> I think we should default to online mode because one of the big drivers for this was not requiring the security escalation for the /etc/hosts modification. 

Note that this has some implications for users on more enterprise-y networks (or home networks that use pfsense or similar). Many routers are now including DNS rebinding protection, which will prevent *.ddev.site from working _at all_.

**Comment by rickmanelius** [comment]
FWIW, @beeradb is out this week and given that he's done a tremendous amount of working weighing in on this, I don't want to move on this until he's back.

**Comment by cweagans** [comment]
Public DNS records that resolve to 127.0.0.1 will be disallowed at the network level on most corporate networks and some home networks using prosumer networking hardware (like mine) via DNS rebinding protection built into the router. Luckily, we can work around this a bit by doing a simple DNS lookup for whatever.ddev.site. If we get `NXDOMAIN`, then there is DNS rebinding protection enabled at the network level and we can fall back to a priv escalation prompt to edit `/etc/hosts`.

Some browsers in some configurations also include DNS rebinding protection - there's not much we can do to detect that or work around it, but it's something to be aware of.

**Comment by tannerjfco** [comment]
I'm closing this as the PR is significantly outdated at this point and I expect somewhat different functionality to come out of the discussion for #416. Further discussion on this issue should be continued there.
